src/app/inventory/store/d2-item-factory.ts

```diff
@@ -201,73 +168,81 @@ export function makeFakeItem(
  * We can use this item to refresh a single item in the store from this response.
  */
 export function makeItemSingle(
-  defs: D2ManifestDefinitions,
-  buckets: InventoryBuckets,
-  item: DestinyItemResponse,
+  context: CreateItemContext,
+  itemResponse: DestinyItemResponse,
   stores: DimStore[]
 ): DimItem | null {
-  if (!item.item.data) {
+  if (!itemResponse.item.data) {
     return null;
   }
 
-  const owner = item.characterId ? stores.find((s) => s.id === item.characterId) : getVault(stores);
+  const owner = itemResponse.characterId
+    ? stores.find((s) => s.id === itemResponse.characterId)
+    : getVault(stores);
 
-  const itemId = item.item.data.itemInstanceId;
+  const itemId = itemResponse.item.data.itemInstanceId;
 
   // Convert a single component response into a dictionary component response
   const empty = { privacy: ComponentPrivacySetting.Public, data: {} };
   const m: <V>(v: SingleComponentResponse<V>) => DictionaryComponentResponse<V> = itemId
     ? (v) => (v ? { privacy: v.privacy, data: v.data ? { [itemId]: v.data } : {} } : empty)
     : () => empty;
 
+  // We'll override our item components with these ones
+  const itemComponents = {
+    instances: m(itemResponse.instance),
+    perks: m(itemResponse.perks),
+    renderData: m(itemResponse.renderData),
+    stats: m(itemResponse.stats),
+    sockets: m(itemResponse.sockets),
+    reusablePlugs: m(itemResponse.reusablePlugs),
+    plugObjectives: m(itemResponse.plugObjectives),
+    talentGrids: m(itemResponse.talentGrid),
+    plugStates: empty,
+    objectives: m(itemResponse.objectives),
+  };
+
   // Make it look like a full response
-  return makeItem(
-    defs,
-    buckets,
-    {
-      instances: m(item.instance),
-      perks: m(item.perks),
-      renderData: m(item.renderData),
-      stats: m(item.stats),
-      sockets: m(item.sockets),
-      reusablePlugs: m(item.reusablePlugs),
-      plugObjectives: m(item.plugObjectives),
-      talentGrids: m(item.talentGrid),
-      plugStates: empty,
-      objectives: m(item.objectives),
-    },
-    item.item.data,
-    owner
-  );
+  return makeItem({ ...context, itemComponents }, itemResponse.item.data, owner);
+}
+
+/**
+ * Stuff that's required to create a DimItem.
+ */
+export interface CreateItemContext {
+  defs: D2ManifestDefinitions;
+  buckets: InventoryBuckets;
+  profileResponse: DestinyProfileResponse;
+  customTotalStatsByClass: {
+    [key: number]: number[];
+  };
+  /**
+   * Sometimes comes from the profile response, but also sometimes from vendors response or mocked out.
+   * If not present, use the one from profileInfo.
```

Comment:
* If not present, the itemComponents from the DestinyProfileResponse should be used